src/charm.py

```diff
@@ -134,7 +133,11 @@ def _configure(self, _):
         # Restart prometheus only if command line arguments have changed,
         # otherwise just reload its configuration.
         if current_services == new_layer.services:
-            reloaded = self._prometheus_server.reload_configuration()
+            external_url = urlparse(self._external_url)
+
+            prometheus_server = Prometheus(web_route_prefix=external_url.path)
```

Comment:
Couple of things I am not clear about (possibly not directly related to this PR)

1. Prometheus [recommends](https://github.com/prometheus/prometheus/issues/4925) not to use both `--web.route-prefix` and `--web.external-url` if the latter **already includes a path**. I do not see any conditional check in our code that ensures this. In fact we are always constructing `_web_route_prefix()` from `_external_url()` and setting it in `_command()` (see `src/charm.py`) . Do we have a potential problem here ? 

2. Do we expect workloads from any related charm may try to make legitimate API requests directly to Prometheus API endpoint ? If so should Prometheus be restarted on an ingress changed event too, so that any new external URL and web route prefix take effect ?

Reply:
@balbirthomas 
1. Where? I recall Prometheus not working correctly without both
2. Well, that's pretty much what `prometheus_remote_write` does, no? Today, if the cli args of prometheus are different, we do restart. It's how https://github.com/canonical/prometheus-k8s-operator/blob/f4fdb6747647af7df93bdea4f8dace581abd30c9/src/charm.py#L145 works AFAICT